### ISSUE TITLE
High CPU utilization

### ISSUE THREAD

**Issue by Bars7878** [wazuh/wazuh#4400]
Hello! After upgrade from wazuh-manager 3.10 to 3.11 watching high load CPU by process ossec-analysisd.
![image](https://user-images.githubusercontent.com/59124940/71562165-83e50f80-2a87-11ea-8b37-5c6b84e240d7.png)

The load does not change and remains constant. OS Ubuntu, 16.04 LTS.

**Comment by Zenidd** [comment]
Hello @Bars7878,

I think it could be due to a large portion of files being monitored through the [ossec-syscheckd](https://documentation.wazuh.com/3.10/user-manual/reference/daemons/ossec-syscheckd.html) daemon (`Syscheck` or `Rootcheck` module). We have a known bug where `Rootcheck` could spend a large portion of time and CPU analyzing some hardcoded paths:

https://github.com/wazuh/wazuh/blob/1c92d178c2e876a774ddd0acf2501e821c2fd688/src/rootcheck/check_rc_sys.c#L365-L372


That's why I would like you to check if disabling the Rootcheck `<check_sys>no</check_sys>` option leads to `ossec-analysisd` using less CPU resources.

```XML
<!-- ossec.conf -->

<rootcheck>
  <check_sys>no</check_sys>
</rootcheck>
```
On the other hand, if you don't want to disable the `<check_sys>` option, you can decrease the priority of the agent service in order to reduce its impact on the system. [Here](https://github.com/wazuh/wazuh/issues/3265#issuecomment-562290021) you have how. 

Let me know how it goes. Greetings, 

JP Sáez

**Comment by Bars7878** [comment]
my configuration:

  <rootcheck>
    <disabled>no</disabled>
    <check_files>yes</check_files>
    <check_trojans>yes</check_trojans>
    <check_dev>yes</check_dev>
    <check_sys>yes</check_sys>
    <check_pids>yes</check_pids>
    <check_ports>yes</check_ports>
    <check_if>yes</check_if>
    
    <frequency>43200</frequency>
    <rootkit_files>/var/ossec/etc/rootcheck/rootkit_files.txt</rootkit_files>
    <rootkit_trojans>/var/ossec/etc/rootcheck/rootkit_trojans.txt</rootkit_trojans>
    <skip_nfs>yes</skip_nfs>
  </rootcheck>

After change <check_sys>yes</check_sys> to <check_sys>no</check_sys> and restart wazuh-manager CPU load don`t change.
Can i debug ossec-anaysisd process and to look which checks give such a result.

**Comment by Zenidd** [comment]
Hello again @Bars7878,

- Could you paste here the `/var/ossec/var/run/ossec-analysisd.state` file content?
`# cat /var/ossec/var/run/ossec-analysisd.state`

- Could you set threads value to 1 alternatively for the different modules and check if it brings down the CPU use? (you can modify this configuration on `/var/ossec/etc/internal_options.conf`)

i.e, setting analysysd n of threads to 1
```
# Number of event decoder threads
analysisd.event_threads=1
# Number of syscheck decoder threads
analysisd.syscheck_threads=0
# Number of syscollector decoder threads
analysisd.syscollector_threads=0
# Number of rootcheck decoder threads
analysisd.rootcheck_threads=0
# Number of security configuration assessment decoder threads
analysisd.sca_threads=0
# Number of hostinfo decoder threads
analysisd.hostinfo_threads=0
# Number of Windows event decoder threads
analysisd.winevt_threads=0
# Number of rule matching threads
analysisd.rule_matching_threads=0
```

I think this is the best way to debug the problem and see what is happening. Please, let me know how it goes.

Greetings, JP Sáez

**Comment by Bars7878** [comment]
Hello! 
```
b@wazuh:~$ sudo cat /var/ossec/var/run/ossec-analysisd.state
# State file for ossec-analysisd
# Total events decoded
total_events_decoded='15'
# Syscheck events decoded
syscheck_events_decoded='0'
syscheck_edps='0'
# Syscollector events decoded
syscollector_events_decoded='0'
syscollector_edps='0'
# Rootcheck events decoded
rootcheck_events_decoded='0'
rootcheck_edps='0'
# Security configuration assessment events decoded
sca_events_decoded='0'
sca_edps='0'
# Hostinfo events decoded
hostinfo_events_decoded='0'
hostinfo_edps='0'
# Winevt events decoded
winevt_events_decoded='0'
winevt_edps='0'
# Other events decoded
other_events_decoded='15'
other_events_edps='3'
# Events processed (Rule matching)
events_processed='15'
events_edps='3'
# Events received
events_received='72'
# Events dropped
events_dropped='57'
# Alerts written to disk
alerts_written='0'
# Firewall alerts written to disk
firewall_written='0'
# FTS alerts written to disk
fts_written='0'
# Syscheck queue
syscheck_queue_usage='0.00'
# Syscheck queue size
syscheck_queue_size='16384'
# Syscollector queue
syscollector_queue_usage='0.00'
# Syscollector queue size
syscollector_queue_size='16384'
# Rootcheck queue
rootcheck_queue_usage='0.00'
# Rootcheck queue size
rootcheck_queue_size='16384'
# Security configuration assessment queue
sca_queue_usage='0.00'
# Security configuration assessment queue size
sca_queue_size='16384'
# Hostinfo queue
hostinfo_queue_usage='0.00'
# Hostinfo queue size
hostinfo_queue_size='16384'
# Winevt queue
winevt_queue_usage='1.00'
# Winevt queue size
winevt_queue_size='16384'
# Event queue
event_queue_usage='0.00'
# Event queue size
event_queue_size='16384'
# Rule matching queue
rule_matching_queue_usage='0.00'
# Rule matching queue size
rule_matching_queue_size='16384'
# Alerts log queue
alerts_queue_usage='0.00'
# Alerts log queue size
alerts_queue_size='16384'
# Firewall log queue
firewall_queue_usage='0.00'
# Firewall log queue size
firewall_queue_size='16384'
# Statistical log queue
statistical_queue_usage='0.00'
# Statistical log queue size
statistical_queue_size='16384'
# Archives log queue
archives_queue_usage='0.00'
# Archives log queue size
archives_queue_size='16384'
b@wazuh:~$ 
```

in my /var/ossec/etc/internal_options.conf:
```
# Number of event decoder threads
analysisd.event_threads=0
# Number of syscheck decoder threads
analysisd.syscheck_threads=0
# Number of syscollector decoder threads
analysisd.syscollector_threads=0
# Number of rootcheck decoder threads
analysisd.rootcheck_threads=0
# Number of security configuration assessment decoder threads
analysisd.sca_threads=0
# Number of hostinfo decoder threads
analysisd.hostinfo_threads=0
# Number of Windows event decoder threads
analysisd.winevt_threads=0
# Number of rule matching threads
analysisd.rule_matching_threads=0
```

Parameter "analysisd.event_threads=0" i change to 1 and restart wazuh-manager. Nothing changed.

**Comment by Zenidd** [comment]
Hello again @Bars7878,

Thanks for the detailed reply!

The `/var/ossec/var/run/ossec-analysisd.state` files gives us some insight of what is happening:
```
# Events received
events_received='72'
# Events dropped
events_dropped='57'
```
- The `Analysisd` module receives the log messages and compares them to the rules. Each kind of event (SCA, FIM, Rootcheck... etc) goes into a related queue waiting for the match stage.
- For 72 events received, `Analysisd` has dropped 57. The `Analysisd` module should be able to handle 8000 events per second without dropping events. 

It seems like at some point, one of the `Analysisd` threads gets stuck with an unexpected Windows event. The thread blocks the CPU, leading to blocked queues and upcoming events dropping. This becomes more noticeable when the `analysisd.winevt_threads=1` as there is only one thread processing and it gets stuck quick.

We are investigating this issue and we think it could be related to this [new formatting for Windows events](https://github.com/wazuh/wazuh/pull/3320). 

Stay tuned to track the fixing process. We will also write any workaround we find or anything that can help with this issue.

Thanks again for your help @Bars7878. Greetings, 
JP Sáez

**Comment by vikman90** [comment]
@Bars7878 this bug has been fixed by #4412.

We will release this patch version early next week.

Thank you very much for reporting this issue. Apologies for the inconvenience.

Kind regards.
